modules/gmc-agent/src/main/java/org/gridgain/action/Session.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gridgain.action;
+
+import org.apache.ignite.internal.processors.authentication.AuthorizationContext;
+import org.apache.ignite.internal.processors.security.SecurityContext;
+import org.apache.ignite.internal.util.typedef.internal.S;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.plugin.security.SecurityCredentials;
+
+import java.net.InetSocketAddress;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Session.
+ */
+public class Session {
+    /** Expiration STATE. It's a final state of lastTouchTime. */
+    private static final Long TIMEDOUT_STATE = 0L;
+
+    /** Session ID. */
+    private final UUID id;
+
+    /** Address. */
+    private InetSocketAddress addr;
+
+    /**
+     * Time when session is used last time.
+     * If this time was set at TIMEDOUT_FLAG, then it should never be changed.
+     */
+    private final AtomicLong lastTouchTime = new AtomicLong(U.currentTimeMillis());
+
+    /** Time when session was invalidated last time. */
+    private final AtomicLong lastInvalidateTime = new AtomicLong(U.currentTimeMillis());
```

Comment:
May be we should use `System.currentTimeMillis()` ?
Because `U.currentTimeMillis()` - updated once in 10 milliseconds in separate thread.

Reply:
Fix.